src/sentry/integrations/slack/unfurl/metric_alerts.py

```diff
@@ -0,0 +1,99 @@
+import html
+import re
+from typing import Any, List, Mapping, Optional
+from urllib.parse import urlparse
+
+from django.db.models import Q
+from django.http.request import HttpRequest, QueryDict
+
+from sentry.incidents.models import AlertRule, Incident, User
+from sentry.integrations.slack.message_builder.metric_alerts import SlackMetricAlertMessageBuilder
+from sentry.models import Integration
+
+from . import Handler, UnfurlableUrl, UnfurledUrl, make_type_coercer
+
+map_incident_args = make_type_coercer(
+    {
+        "org_slug": str,
+        "alert_rule_id": int,
+        "incident_id": int,
+    }
+)
+
+
+def unfurl_metric_alerts(
+    request: HttpRequest,
+    integration: Integration,
+    links: List[UnfurlableUrl],
+    user: Optional["User"] = None,
+) -> UnfurledUrl:
+    alert_filter_query = Q()
+    incident_filter_query = Q()
+    # Since we don't have real ids here, we use the org slug so that we can
+    # make sure the identifiers correspond to the correct organization.
+    for link in links:
+        org_slug = link.args["org_slug"]
+        alert_rule_id = link.args["alert_rule_id"]
+        incident_id = link.args["incident_id"]
+
+        if incident_id:
+            incident_filter_query |= Q(
+                identifier=incident_id, organization__slug=org_slug, alert_rule__id=alert_rule_id
+            )
+        else:
+            alert_filter_query |= Q(id=alert_rule_id, organization__slug=org_slug)
+
+    all_integration_orgs = integration.organizations.all()
+    alert_rule_map = {
+        rule.id: rule
+        for rule in AlertRule.objects.filter(
+            alert_filter_query,
+            # Filter by integration organization here as well to make sure that
+            # we have permission to access these incidents.
+            organization__in=all_integration_orgs,
```

Comment:
Ok this looks crazy but I swear it works:
```py
alert_rule_map = {
    rule.id: rule
    for rule in AlertRule.objects.filter(
        alert_filter_query,
        organization__organizationintegration__integration=integration,
    )
}
incident_map = {
    i.identifier: i
    for i in Incident.objects.filter(
        incident_filter_query,
        organization__organizationintegration__integration=integration,
    )
}
```
This will let you omit the `integration.organizations.all()` call.
Also you can move:
```py
if not alert_rule_map:
    return {}
```
up so that you can short-circuit before making the `Incident` DB call.

Reply:
It also might be clearer to do:
```py
alert_filter_query = Q(organization__organizationintegration__integration=integration)
incident_filter_query = Q(organization__organizationintegration__integration=integration)
...
alert_rules = AlertRule.objects.filter(alert_filter_query)
if not alert_rules:
    return {}

incidents = Incident.objects.filter(incident_filter_query)
alert_rule_map = {rule.id: rule for rule in alert_rules}
incident_map = {i.identifier: i for i in incidents}
```

Reply:
These actually make roughly equivalent queries. The current method will do 
```
select * 
from sentry_alertrule
where <other_conditions> 
AND "sentry_alertrule"."organization_id" IN
(
        SELECT id
        FROM sentry_organization
        INNER JOIN "entry_organizationintegrationON (sentry_organization.id = sentry_organizationintegration.organization_id)
        WHERE integration_id = 1
)
``` 
and the new query will do something similar but with joins instead. The join method can cause row duplication, but since we have a unique constraint on `integration, org_id` it should be fine.

Since django querysets are lazy, either way we're not making extra queries, so it's up to personal preference.

Reply:
🤷‍♂️